File: DAMOS/trunk/tests/org.eclipselabs.damos.codegen.c.tests/src/org/eclipselabs/damos/codegen/c/gtests/UnionTest.java
/****************************************************************************
 * Copyright (c) 2008, 2012 Andreas Unger and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *    Andreas Unger - initial API and implementation 
 ****************************************************************************/

package org.eclipselabs.damos.codegen.c.gtests;

import org.eclipselabs.damos.codegen.c.test.AbstractGeneratorGTest;
import org.eclipselabs.damos.codegen.c.test.GTest;
import org.eclipselabs.damos.dml.Block;
import org.eclipselabs.damos.dml.Inport;
import org.eclipselabs.damos.dml.Outport;
import org.junit.Before;

/**
 * @author Andreas Unger
 *
 */
@GTest(sourceFile="gtests/UnionTest.cpp", program="gtests/UnionTest")
public class UnionTest extends AbstractGeneratorGTest {

	@Before
	@Override
	public void setUp() {
		super.setUp();
		
		Inport inA = createInport("InA", createIntegerTypeSpecification(), 1);
		Inport inB = createInport("InB", createIntegerTypeSpecification(), 1);
		Block unionTest = createTestBlock("UnionTest", "UnionTest");
		
		Block inspectTest1 = createTestBlock("InspectTest", "InspectTest1");
		Block inspectTest2 = createTestBlock("InspectTest", "InspectTest2");

		Outport outport1 = createOutport("Out1", createDataTypeSpecification("union { int x; real y }"));
		Outport outport2 = createOutport("Out2", createDataTypeSpecification("union { int x; real y }"));
		
		connect(inA, unionTest, 0);
		connect(inB, unionTest, 1);
		connect(unionTest, 0, inspectTest1);
		connect(unionTest, 1, inspectTest2);
		connect(inspectTest1, outport1);
		connect(inspectTest2, outport2);

		generateAndCompile();
	}

}

File: DAMOS/trunk/tests/org.eclipse.damos.codegen.c.tests/gtests/UnionTest.cpp
/*******************************************************************************
 * Copyright (c) 2008, 2012 Andreas Unger and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *     Andreas Unger - initial API and implementation
 *******************************************************************************/
#include "gtest/gtest.h"
#include "UnionTest.h"

TEST(UnionTest, Union) {
	Input input;
	Output output;

	initialize();

	input.inA = 3;
	input.inB = 7;

	execute(&input, &output);

	EXPECT_EQ(0, output.out1.tag);
	EXPECT_EQ(30, output.out1.value.x);

	EXPECT_EQ(1, output.out2.tag);
	EXPECT_DOUBLE_EQ(140.0, output.out2.value.y);
}
